### PR TITLE
TMA-532: Rollout SynchronizeLdm: improve behavior of update_preferences

### PR DESCRIPTION
**Pull request by vhtien**
## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog
- [ ] Build passing

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/824/ : SUCCESS in 37s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/780/ : FAILURE in 3m 16s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/552/ : FAILURE in 29s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/825/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/781/ : FAILURE in 3m 00s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/553/ : FAILURE in 25s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/826/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-tests-zuul/782/ : SUCCESS in 45m 16s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/554/ : FAILURE in 20s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/5/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/839/ : SUCCESS in 17s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/38/ : SUCCESS in 5s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/862/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/9/ : SUCCESS in 2m 48s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/47/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/885/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/32/ : FAILURE in 5s

**Comment by vhtien** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/52/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/890/ : SUCCESS in 29s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/37/ : SUCCESS in 21s

File: spec/unit/models/project_creator_spec.rb [hunk 1]
```diff
@@ -71,4 +71,19 @@
       GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { cascade_drops: true, preserve_data: false, unmeetable_condition: true })
     end.to raise_error
   end
+
+  it 'should pick correct update chunks based on new parameter "allow_cascade_drops" in your preference' do
+    chunk = GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { allow_cascade_drops: false })
+    expect(chunk).to eq [@chunk_a]
```
Comment (panjan) [inline spec/unit/models/project_creator_spec.rb]: I would expect `[@chunk_a, @chunk_b]`.

File: spec/unit/models/project_creator_spec.rb [hunk 2]
```diff
@@ -71,4 +71,19 @@
...
+  end
+
+  it 'should pick correct update chunks based on new parameter "keep_data" in your preference' do
+    chunk = GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { keep_data: false })
+    expect(chunk).to eq [@chunk_a, @chunk_b]
+  end
+
+  it 'should use new parameters in your preference instead of the old ones' do
+    chunk = GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { allow_cascade_drops: true, cascade_drops: false })
```
Comment (panjan) [inline spec/unit/models/project_creator_spec.rb]: `allow_cascade_drops: true` contradicts  `cascade_drops: false`. IMHO, mixing the original parameters (cascade_drops, preserve_data) with new ones (allow_ cascade_drops, keep_data) should result in an error.

File: spec/unit/models/project_creator_spec.rb [hunk 3]
```diff
@@ -71,4 +71,19 @@
...
   end
+
+  it 'should pick correct update chunks based on new parameter "allow_cascade_drops" in your preference' do
+    chunk = GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { allow_cascade_drops: false })
+    expect(chunk).to eq [@chunk_a]
+  end
+
+  it 'should pick correct update chunks based on new parameter "keep_data" in your preference' do
+    chunk = GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { keep_data: false })
+    expect(chunk).to eq [@chunk_a, @chunk_b]
```
Comment (panjan) [inline spec/unit/models/project_creator_spec.rb]: I would expect `[@chunk_b, @chunk_d]`.

File: spec/unit/models/project_creator_spec.rb (hunk 1 above)
Comment (vhtien) [inline spec/unit/models/project_creator_spec.rb]: It's not correct. Maybe you forgot the default values.

File: spec/unit/models/project_creator_spec.rb (hunk 3 above)
Comment (vhtien) [inline spec/unit/models/project_creator_spec.rb]: It's not correct. Maybe you forgot the default values.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/72/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/912/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/58/ : SUCCESS in 19s

File: spec/unit/models/project_creator_spec.rb (hunk 2 above)
Comment (vhtien) [inline spec/unit/models/project_creator_spec.rb]: done

File: spec/unit/models/project_creator_spec.rb [hunk 4]
```diff
@@ -71,4 +71,20 @@
...
+  end
+
+  it 'should raise error when mixing new parameters with the old ones in your preference' do
+    expect do
+      GoodData::Model::ProjectCreator.pick_correct_chunks(@data, update_preference: { allow_cascade_drops: true, cascade_drops: false })
+    end.to raise_error
```
Comment (panjan) [inline spec/unit/models/project_creator_spec.rb]: Please check that _the right_ error was raised by passing a parameter to `raise_error`.

Comment (vhtien) [inline spec/unit/models/project_creator_spec.rb]: done

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/74/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/914/ : SUCCESS in 37s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/60/ : SUCCESS in 24s

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/34/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/920/ : SUCCESS in 17s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/66/ : SUCCESS in 23s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/843/ : SUCCESS in 44m 18s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/616/ : UNSTABLE in 19m 01s

**Comment by panjan** [comment]
It looks like the build failure is legit. @vhtien

**Comment by vhtien** [comment]
Rebuild the failed test is ok. https://ci.intgdc.com/job/appstore-lcm-integration-tests-zuul/324/console. So please merge this PR again

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/36/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-rubocop-zuul/927/ : SUCCESS in 20s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/73/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/846/ : SUCCESS in 45m 14s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/619/ : SUCCESS in 23m 27s

**Pull request merged**